### ISSUE TITLE
Unhandled exception from twisted/protocols/tls.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5439]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5439|
|Type|release blocker: regression|
|Created|2011-12-22 02:45:41Z|
|Branch|https://github.com/twisted/twisted/tree/tls-error-5439|

Seen on Twisted 11.1.0:

```

2011-12-21 19:32:16-0500 [Broker,client] Unhandled Error
        Traceback (most recent call last):
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/banana.py", line 153, in gotItem
            self.callExpressionReceived(item)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/banana.py", line 116, in callExpressionReceived
            self.expressionReceived(obj)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/pb.py", line 584, in expressionReceived
            method(*sexp[1:])
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/pb.py", line 896, in proto_message
            self._recvMessage(self.localObjectForID, requestID, objectID, message, answerRequired, netArgs, netKw)
        --- <exception caught here> ---
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/pb.py", line 910, in _recvMessage
            netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/spread/flavors.py", line 213, in remoteMessageReceived
            state = apply(method, (self.perspective,)+args, kw)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/web/server.py", line 298, in view_registerProducer
            self.registerProducer(_RemoteProducerWrapper(producer), streaming)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/web/http.py", line 804, in registerProducer
            self.transport.registerProducer(producer, streaming)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/protocols/tls.py", line 547, in registerProducer
            self.transport.registerProducer(producer, True)
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/internet/abstract.py", line 90, in registerProducer
            producer.stopProducing()
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/protocols/tls.py", line 204, in stopProducing
            self._producer.stopProducing()
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/protocols/tls.py", line 155, in stopProducing
            self.stopStreaming()
          File "/srv/pypy/pypy-c-jit-43780-b590cf6de419-linux/site-packages/twisted/protocols/tls.py", line 142, in stopStreaming
            self._coopTask.stop()
        exceptions.AttributeError: '_PullToPush' object has no attribute '_coopTask'
```

It looks like the `registerProducer` case where the consumer is a transport which is already disconnected does something that the new TLS producer/consumer code isn't too happy with - calls `stopProducing` on it, even though `startProducing` has never been called.

This is a bit unsightly, but fortunately it doesn't seem to be a very severe problem, since it must be happening to connections that have been closed already (hence no data is lost).

<details><summary>Searchable metadata</summary>

```
trac-id__5439 5439
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__normal normal
milestone__Twisted_12_0 Twisted-12.0
branch__branches_tls_error_5439 branches/tls-error-5439
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1324521941000000 1324521941000000
changetime__1327248239000000 1327248239000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [33460]) Branching to 'tls-error-5439'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst removed owner|
|-|-|

Ready for review.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tls-error-5439

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. The docstring for `test_streamingProducerAfterConnectionLost` has a mistake, it says _non-streaming_ where it means _streaming_.
  1. The news fragment doesn't follow the style recommendations on ReviewProcess#Newsfiles.

Otherwise good.  Please merge when the above are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [33470]) Merge tls-error-5439.

Author: itamar
Review: exarkun
Fixes: [#5439](https://github.com/twisted/twisted/issues/5439)

If a producer is registered on a TLS transport after connection lost, just stop it immediately.
